assets/js/base/context/cart-checkout/payment-methods/payment-method-data-context.js

```diff
@@ -95,9 +95,17 @@ const getCustomerPaymentMethods = ( availablePaymentMethods = [] ) => {
 	paymentMethodKeys.forEach( ( type ) => {
 		const methods = customerPaymentMethods[ type ].filter(
 			( paymentMethod ) => {
-				return Object.keys( availablePaymentMethods ).includes(
-					paymentMethod.method.gateway
-				);
+				const { gateway } = paymentMethod.method;
+				const isGatewayAvailable = Object.keys(
+					availablePaymentMethods
+				).includes( gateway );
+				if ( ! isGatewayAvailable ) {
+					return false;
+				}
+				const supportsSavedPayments =
+					availablePaymentMethods[ gateway ].supports
+						?.savePaymentInfo;
+				return supportsSavedPayments;
```

Comment:
Based on your changes, I wonder if we could simplify this down if the gateway exists in the list of `availablePaymentMethods`

```suggestion
				return (
					typeof availablePaymentMethods[
						paymentMethod.method.gateway
					] !== 'undefined' &&
					availablePaymentMethods[ paymentMethod.method.gateway ]
						.supports?.savePaymentInfo
				);
```

Reply:
We could, but I would like to avoid it in this case. A agree that the brevity of code is a good quality in general. The context and deeply embedded nature of this suggest that it should be painfully obvious what this does for the future. This is the second time in one month I am revisiting the logic and adding additional checks. I bet that this is not the last time we will do updates here.

Reply:
Hmm, I found the original code hard to understand to be honest.

Maybe you could play around with it slightly, for example:

```
const gateway = paymentMethod.method.gateway;
const isGatewayAvailable = typeof availablePaymentMethods[ gateway ] !== 'undefined';
const supported = isGatewayAvailable ? availablePaymentMethods[ gateway ].supports?.savePaymentInfo : false;
return supported === true;
```

Still more compact but easier to read.

Reply:
OK I will take a look.

Reply:
I did a bit of research on how to approach this in a most idiomatic way. For the check, if the gateway exists I have used the `in` operator. It feels like it was designed with cases like this in mind
```js
const isAvailable = gateway in availablePaymentMethods;
```
For the second test I have used this simple construct:
```js
return (
	isAvailable &&
	availablePaymentMethods[ gateway ].supports?.savePaymentInfo
);
```
I have replaced the `?` in your suggestions with `&&` to shortcircuit the evaluation. We don't need to worry about the potential `undefined` from the optional chaining operator. The returned value is used by `Array.prototype.filter` and by definition in callback description, we have `Return a value that coerces to true to keep the element, or to false otherwise.` Since `undefined` coerces to false it is a good way to indicate that we don't want to include that gateway method.
@mikejolley thanks for the pushback. It resulted in a much better and cleaner code.

As for the parameter type issue: I am creating a separate PR to fix that, there will be one additional change that would clutter the scope and sense of this PR.